src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasGoogleCloudService.kt

```diff
@@ -0,0 +1,84 @@
+
+import com.google.cloud.storage.*
+import io.grpc.stub.StreamObserver
+import com.google.cloud.storage.Acl.User
+import com.google.cloud.storage.contrib.nio.testing.LocalStorageHelper
+import com.google.protobuf.ByteString
+import io.grpc.Status
+import io.grpc.StatusException
+import java.util.Arrays
+import java.util.ArrayList
+
+/**
+ * Special class that can work with requests from CosmasClient
+ * This realization stores files in Google Cloud Storage.
+ *
+ * @author Aleksandr Fedotov (iisuslik43)
+ */
+class CosmasGoogleCloudService(private val bucketName: String) : CosmasGrpc.CosmasImplBase() {
+
+    private var storage: Storage = StorageOptions.getDefaultInstance().service
+
+    override fun createVersion(request: CosmasProto.CreateVersionRequest,
+                               responseObserver: StreamObserver<CosmasProto.CreateVersionResponse>) {
+        println("Get request for create new version of file № ${request.fileId}")
+        storage.create(
+                BlobInfo.newBuilder(bucketName, request.fileId)
+                        // Modify access list to allow all users with link to read file
+                        .setAcl(ArrayList(Arrays.asList(Acl.of(User.ofAllUsers(), Acl.Role.READER))))
+                        .build(),
+                request.file.toByteArray())
+        val response: CosmasProto.CreateVersionResponse = CosmasProto.CreateVersionResponse
+                .newBuilder()
+                .build()
+        responseObserver.onNext(response)
+        responseObserver.onCompleted()
+    }
+
+    override fun getVersion(request: CosmasProto.GetVersionRequest,
+                            responseObserver: StreamObserver<CosmasProto.GetVersionResponse>) {
+        println("Get request for version ${request.version} file № ${request.fileId}")
+        val blob = storage.get(BlobInfo.newBuilder(bucketName, request.fileId).build().blobId)
+        val response = CosmasProto.GetVersionResponse.newBuilder()
+        if (blob != null) {
+            response.file = ByteString.copyFrom(blob.getContent())
+        } else {
+            val requestStatus = Status.NOT_FOUND.withDescription("There is no such file in storage")
+            println("This request is incorrect: " + requestStatus.description)
+            responseObserver.onError(StatusException(requestStatus))
+        }
+        responseObserver.onNext(response.build())
+        responseObserver.onCompleted()
+    }
+
+    fun deleteFile(fileId: String) {
+        println("Delete file № $fileId")
+        storage.delete(BlobInfo.newBuilder(bucketName, fileId).build().blobId)
+    }
+
+    companion object {
+        fun getServiceForTests(): CosmasGoogleCloudService {
+            return CosmasGoogleCloudService("papeeria-interns-cosmas",
+                    LocalStorageHelper.getOptions().service)
+        }
+    }
+
+    private constructor(bucketName: String, testStorage: Storage) : this(bucketName) {
```

Comment:
Please move this upwards, after member fields declaration. Constructors are important guys and deserve being placed the first.

Another option is to add storage argument to the primary constructor and provide a default value (the real storage)

Reply:
Ok, but I like private constructor

Reply:
Private constructors are nice when they have some good reasons behind. Do not allow anyone to create instances of this class (and restrict to a set of predefined instances) is a good reason. But here you actually do need to parameterize this class with different `Storage` implementations. If you start writing tests for the cases when `Storage` throws exceptions (and you normally want such tests, because they do throw exceptions) you will find our that throwing exception from the real or in-memory implementation is not that easy. The easiest way is using mocks. And the only way to pass mock implementation from tests is to allow for passing `Storage` instance.

In general, usually if your code depends on some non-trivial component, it is a good idea to parameterize your code with instance of that component interface, rather than access the component instance from the code itself. Parameterizing makes your code more test-friendly and more change-friendly: you explicitly declare the interface you depend on instead of hiding such interface in the class internals.